grails-app/services/org/pih/warehouse/inventory/ProductAvailabilityService.groovy

```diff
@@ -155,33 +195,77 @@ class ProductAvailabilityService {
+        }
+        def picked = getQuantityPickedByProductAndLocation(location, product)
+        return transformBinLocations(binLocationsWithQuantity, picked)
+    }
+
+    /**
+     * Get quantity allocated for bin locations and inventory items for a specific location (origin) and product.
+     * Data is pulled from:
+     *  1. Picklist items from pending requisitions (outbound stock movements) with origin being provided location
+     *     that are *NOT* having RECALLED inventory item (inventoryItem.lotStatus) or bin location *WITHOUT*
+     *     HOLD_STOCK in the supported activity (location.supportedActivities).
+     *  2. Picklist items from pending orders (outbound returns) with origin being provided location.
+     *     (IMPORTANT: Outbound returns can have picked items with RECALLED lots and bins with HOLD_STOCK activity)
+     * */
     def getQuantityPickedByProductAndLocation(Location location, Product product) {
-        def results = Picklist.executeQuery("""
-            SELECT
-                pli.binLocation,
-                pli.inventoryItem,
-                sum(pli.quantity)*((count(distinct pli.requisitionItem) + count(distinct pli.orderItem) )/ count(*))
-            FROM PicklistItem pli
-            INNER JOIN pli.picklist pl
-            LEFT JOIN pli.requisitionItem ri
-            LEFT JOIN pl.requisition r
-            LEFT JOIN pli.orderItem oi
-            LEFT JOIN pl.order o
-            LEFT JOIN pli.inventoryItem ii
-            LEFT JOIN pli.binLocation l
-            LEFT JOIN l.supportedActivities s
-            WHERE ((r.origin = :location AND r.status IN (:pendingRequisitionStatus)) OR (o.origin = :location AND o.status IN (:pendingOrderStatus)))
-              AND (oi IS NOT NULL OR (ri IS NOT NULL AND (ii.lotStatus IS NULL OR ii.lotStatus != 'RECALLED') AND NOT ('HOLD_STOCK' IN ELEMENTS(l.supportedActivities))))
-              AND (:product = '' OR ri.product.id = :product OR oi.product.id = :product)
-            GROUP BY pli.binLocation, pli.inventoryItem
-        """, [
-                location                    : location,
-                pendingRequisitionStatus    : RequisitionStatus.listPending(),
-                pendingOrderStatus          : OrderStatus.listPending(),
-                product                     : product?.id ?: ''
-        ], [readOnly: true])
-
-        return results.collect { [binLocation: it[0]?.id, inventoryItem: it[1]?.id, quantityAllocated: it[2]] }
+        def query = """
+            SELECT bin_location_id, inventory_item_id, SUM(quantity) FROM (
+                SELECT
+                    pli.bin_location_id as bin_location_id,
+                    pli.inventory_item_id as inventory_item_id,
+                    sum(pli.quantity) as quantity
+                FROM picklist_item pli
+                    INNER JOIN picklist p ON pli.picklist_id = p.id
+                    LEFT JOIN requisition_item ri ON pli.requisition_item_id = ri.id
```

Comment:
In my opinion, we don't need to use LEFT JOIN any longer. That was one of the reasons why I wanted to do a UNION here. I think location (bin location) is the only table that needs to be LEFT JOIN'd.